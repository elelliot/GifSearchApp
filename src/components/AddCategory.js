import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory =  ( {setCategories} ) => {

    const [inputValue, setInputValue] = useState('');//string vacío, si lo dejamos sin '', es undefined y causará errores

    const handleInputChange = ( e ) => {
        //console.log(e.target.value);
        setInputValue (e.target.value);//obtenemos los cambios de texto con el evento

        //console.log('handleInputChange llamado');
    }

    const handleSubmit = ( e ) =>{

        e.preventDefault()//prevenimos el comportamiento por defecto que sería refrescar
        // console.log('Submit Hecho')

        //console.log('handleSubmit', inputValue);


        //validamos que no haya valores nulos o que sean mínimo de 3 caracteres
        if( inputValue.trim().length > 2){
            
            /*como no pasé las categories como argumento junto con el setCategories 
            llamo al callback que tiene el estado anterior, en este caso le llamo cats*/
            setCategories( cats => [ inputValue, ...cats]);//insertamos primero la categoría que buscamos y despues las anteriores
            setInputValue('');
        }


    }


    return (
        //cuando se quiere regresar un formulario, los fragments son redundantes ya que el form agrupa o el div
        //cuando aplastas enter el form se envía y se refresca por defecto
        <form onSubmit = { handleSubmit }>
            <p> { inputValue }</p>
            <input 
                type="text"
            /*el value se queda igual ya que esto se redibuja cada que toco el teclado y agarramos el valor desde el evento onChange*/
               value= { inputValue }
            /*Se puede mandar el primer argumento del onChange que sería el evento y lo atrapamos en la función (arriba) */
                onChange = { handleInputChange /*e => console.log(e)*/ } 
            />
        </form>
        
    )
}
//hacemos que el setCategories sea funcion obligatoria
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
