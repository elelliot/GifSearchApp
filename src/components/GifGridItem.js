import React from 'react'
import PropTypes from 'prop-types';

export const GifGridItem = ( { /*id,*/ title, url } ) => {

    // console.log( id, title, url  );

    return (
        <div className="card animate__animated animate__fadeIn ">
            <img src= { url } alt = { title } />
            <p>{ title }</p>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
}
// import React from 'react'

// export const GifGridItem = ( img ) => {

//     console.log( img );
    
//     return (
//         <div>
//             { img.title }
//         </div>
//     )
// }

















