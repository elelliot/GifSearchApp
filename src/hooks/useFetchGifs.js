import { useEffect, useState } from "react"
import { getGifs } from '../helpers/getGifs'

export const useFetchGifs = ( category ) => {

    const [state, setState] = useState({
        data: [],
        loading: true,

    });


    useEffect( () =>{

        getGifs( category )
            .then( imgs =>{

                setState( {
                    data: imgs,
                    loading: false
                });
            })

    },[category])


    

    return state; //{ data: [] , loading: true }
}











/*el use effect recibe una funcion , la que quiero ejecutar,
     y gracias al arreglo de dependencias vacío solo se dispara una única vez
    */
    // useEffect( ()=> {
    //     //solo quiero que se ejecute esta instruccion cuando el componente es renderizado por 1ra vez
    //     //como cambia la categoría queremos volver a ejecutarla, entonces la ponemos dentro del argumento del array
    //     getGifs( category )
    //         .then( setImages ); //como tenemos una funcion cuyo primer arg es enviado como 1er arg podemos cambiar imgs => setImages( imgs ) a lo que tenemos
    // }, [ category ])




    // setTimeout( () => {
    //     setState({
    //         data: [1,2,3,4,5,6,7],
    //         loading: false,
    
    //     })
    // }, 3000);