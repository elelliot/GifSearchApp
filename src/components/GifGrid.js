import React from 'react';
import Proptypes from 'prop-types';

import { useFetchGifs } from '../hooks/useFetchGifs'
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    //usamos el custom hook
    const { data:images , loading } = useFetchGifs( category ); //renombramos la data desestructurada


    // const [images, setImages] = useState([])


    return (
        <>
            <h3 className=" animate__animated animate__fadeIn ">{ category }</h3>

            {/* para no tener que poner un null si la condicion no se cumple usamos el && para solo evaluar lo primero 
            y no hacer nada si no se cumple */}
            { loading && <p className = "animate__animated animate__flash">Loading</p> }
            {/*  loading ? <p>Loading</p> : null */}
            
            <div className = "card-grid">
                
                {
                    images.map( img => (
                        <GifGridItem 
                            key = { img.id }
                            { ...img } 
                            // img = { img } 
                        />
                        // return <li key= { imagen.id }>{ imagen.title }</li>
                    ))
                    
                }
                
            </div>
        </>
    )
}


GifGrid.propTypes = {
    category: Proptypes.string.isRequired
}

