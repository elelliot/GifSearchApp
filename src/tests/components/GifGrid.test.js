import { GifGrid } from "../../components/GifGrid"
import { shallow } from "enzyme"
import '@testing-library/jest-dom'
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock('../../hooks/useFetchGifs'); //fingimos la llamada a ese archivo


describe('Pruebas en <GifGrid />', () => {
    

    const category = 'Samurai X';


    //Mock Custom Hooks
    test('Debe mostrar el componente correctamente (El mensaje de Loading) ', () => {

        //falsa implementacion de cuando no tenemos una respuesta
        useFetchGifs.mockImplementation( category => { 
            return { data: [], loading: true }
          });

        const wrapper = shallow( <GifGrid category = { category } /> );
        //El Snapshot muestra el Loading unicamente, no los items
        expect( wrapper ).toMatchSnapshot();
    });


    test('Debe mostrar items cuando se cargan imagenes useFetchGifs ', () => {
        
        //evaluar cuando ya tienes una respuesta
        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'Cualquier cosa'
        },
        {
            id: '123',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'Cualquier cosa'
        }] 


        useFetchGifs.mockImplementation( category => {
            return { data: gifs, loading: false };
          });

        const wrapper = shallow( <GifGrid category = { category } /> );


        //expect( wrapper ).toMatchSnapshot();

        //si el loading está en false, el parrafo no debería aparecer
        expect( wrapper.find('p').exists() ).toBe(false);

        //Los GifGridItem de la coleccion deben ser del mismo tamaño del numero de gifs que recibo 2 y 2 por ejemplo
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );


    })
    
    
})
