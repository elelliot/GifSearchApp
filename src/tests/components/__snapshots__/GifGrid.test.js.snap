// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> Debe mostrar el componente correctamente (El mensaje de Loading)  1`] = `
<Fragment>
  <h3
    className=" animate__animated animate__fadeIn "
  >
    Samurai X
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
